### PR TITLE
Integrate Clippy into the project

### PR DESCRIPTION
**Pull request by blyxyas**
This PR integrates Clippy into the project.
A user can run the `clippy` profile with a local Clippy build, with the `--clippy` flag targetting to the Clippy executable:

```sh
$ cargo run --bin collector  -- bench_local +nightly --profiles Clippy --clippy ../clippy/target/release/cargo-clippy
```

---

But I'm having some problems with the bot integration, so some feedback and tips in those files would be very appreciated.

* Would the bot run this on every commit, where is it configured? (we don't want to run it every commit)
Thanks for the feedback :heart:

### CODE REVIEW

File: collector/compile-benchmarks/cargo-0.60.0/src/cargo/core/profiles.rs [hunk 1]
```diff
@@ -95,6 +95,14 @@ impl Profiles {
                 .dir_names
                 .insert(InternedString::new("doc"), InternedString::new("debug"));
 
+            profile_makers.by_name.insert(
```
Comment (Kobzol) [inline collector/compile-benchmarks/cargo-0.60.0/src/cargo/core/profiles.rs:98]: Why did you need to change the source of the cargo benchmark? It shouldn't be required to change a benchmark to run `clippy` on it.

File: collector/src/bin/collector.rs [hunk 2]
```diff
@@ -663,6 +667,7 @@ fn main_result() -> anyhow::Result<i32> {
                     None,
```
Comment (Kobzol) [inline collector/src/bin/collector.rs:667]: I think that we have gotten past the point where N `None`s in a row is an acceptable API :laughing: I know that it's mostly preexisting, but would you be willing to turn this e.g. into a builder pattern? Something like
```rust
get_local_toolchain(..., ToolchainConfig::default()) // everything `None` by default
get_local_toolchain(..., ToolchainConfig::default().rustc(...).clippy(...)) // some things filled
```
it could contain all parameters for the `get_local_toolchain` function, where it makes sense.

File: collector/src/compile/benchmark/profile.rs [hunk 3]
```diff
@@ -12,13 +12,15 @@ pub enum Profile {
     Debug,
     Doc,
     Opt,
+    Clippy,
 }
 
 impl Profile {
     pub fn all() -> Vec<Self> {
```
Comment (Kobzol) [inline collector/src/compile/benchmark/profile.rs:19]: This will need to be modified in `BenchNext` and `bench_published_artifact`. These two functions should not benchmark Clippy. Probably the functions should directly decide the specific profiles to benchmark, rather than delegating to these two functions (`all` and `all_non_doc`).

File: collector/src/toolchain.rs [hunk 4]
```diff
@@ -400,6 +405,26 @@ pub fn get_local_toolchain(
             None
         };
 
+        let clippy =
+        if let Some(clippy) = &clippy {
+            Some(clippy.canonicalize().with_context(|| {
+                format!("failed to canonicalize rustdoc executable {:?}", clippy)
```
Comment (Kobzol) [inline collector/src/toolchain.rs:411]:
```suggestion
                format!("failed to canonicalize clippy executable {:?}", clippy)
```

File: collector/src/toolchain.rs [hunk 5]
```diff
@@ -400,6 +405,26 @@ pub fn get_local_toolchain(
...
+            })?)
+        } else if profiles.contains(&Profile::Clippy) {
+            // We need a `clippy`. Look for one next to `rustc`.
+            if let Ok(clippy) = rustc.with_file_name("clippy").canonicalize() {
```
Comment (Kobzol) [inline collector/src/toolchain.rs:415]: I think that this should be `cargo-clippy` instead of `clippy`. It is called `cargo-clippy` in my local toolchains installed by `rustup`.

File: collector/src/toolchain.rs [hunk 6]
```diff
@@ -400,6 +405,26 @@ pub fn get_local_toolchain(
...
+                debug!("found clippy: {:?}", &clippy);
+                Some(clippy)
+            } else {
+                anyhow::bail!(
+                    "'Clippy' build specified but '--clippy' not specified and no 'clippy' found \
```
Comment (Kobzol) [inline collector/src/toolchain.rs:420]:
```suggestion
                    "'Clippy' build specified but '--clippy' not specified and no 'cargo-clippy' found \
```

File: site/src/request_handlers/github.rs [hunk 7]
```diff
@@ -123,8 +123,15 @@ async fn handle_rust_timer(
         let exclude = captures.get(2).map(|v| v.as_str());
         let runs = captures.get(3).and_then(|v| v.as_str().parse::<i32>().ok());
         {
+            let profile;
```
Comment (Kobzol) [inline site/src/request_handlers/github.rs:126]: I'm not sure about the integration of clippy into the main dashboard. I'd rather leave this part of the PR out for now, and just focus on data gathering.

File: database/src/pool/sqlite.rs [hunk 8]
```diff
@@ -957,13 +957,14 @@ impl Connection for SqliteConnection {
         include: Option<&str>,
         exclude: Option<&str>,
         runs: Option<i32>,
+        profile: Option<&str>
     ) {
         self.raw_ref()
             .prepare_cached(
-                "insert into pull_request_build (pr, complete, requested, include, exclude, runs) VALUES (?, 0, strftime('%s','now'), ?, ?, ?)",
+                "insert into pull_request_build (pr, complete, requested, include, exclude, runs, profile) VALUES (?, 0, strftime('%s','now'), ?, ?, ?, ?)",
```
Comment (Kobzol) [inline database/src/pool/sqlite.rs:964]: Why did you change SQL queries? I don't think that we need to change anything SQL related in this PR. Also, if we did, we would have to write a migration.

**Review by Kobzol — COMMENTED** [review]
Hi, thanks for the PR! I left some comments. I wonder what was your motivation for this PR? Do you work on `Clippy` and want to measure its performance on crates?

**Comment by blyxyas** [comment]
> I wonder what was your motivation for this PR? Do you work on Clippy and want to measure its performance on crates?
Yep, we wanted a tool for measuring Clippy's performance on crates.

File: site/src/request_handlers/github.rs (hunk 7 above)
Comment (blyxyas) [reply site/src/request_handlers/github.rs:126]: Would the data still be viewable outside the dashboard somewhere?

Comment (Kobzol) [reply site/src/request_handlers/github.rs:126]: It would be visible only in the local DB. I would like to first merge support for benchmarking clippy. We can discuss its inclusion in the server after that :)

**Comment by blyxyas** [comment]
This new commit should be just like the first one, but discarding anything server-related =^w^=

File: collector/src/toolchain.rs [hunk 9]
```diff
@@ -123,6 +123,7 @@ impl SysrootDownload {
         let components = ToolchainComponents::from_binaries_and_libdir(
             sysroot_bin("rustc")?,
             Some(sysroot_bin("rustdoc")?),
+            Some(sysroot_bin("clippy")?),
```
Comment (Kobzol) [inline collector/src/toolchain.rs:126]: As is apparent on CI, `clippy` might not always be applicable. So we should only pass it here if it actually exists. Also, it should also probably be `cargo-clippy` here.

File: collector/src/toolchain.rs [hunk 10]
```diff
@@ -288,6 +292,45 @@ impl ToolchainComponents {
...
+#[derive(Clone, Copy)]
+pub struct ToolchainConfig<'a> {
+    rustdoc: Option<&'a Path>,
+    clippy: Option<&'a Path>,
+    cargo: Option<&'a Path>,
+    id: Option<&'a str>,
+}
+
+impl<'a> ToolchainConfig<'a> {
+    pub fn default() -> Self {
```
Comment (Kobzol) [inline collector/src/toolchain.rs:304]: Somewhat ironically, clippy tells you to derive Default for this struct :laughing: https://github.com/rust-lang/rustc-perf/actions/runs/6273536371/job/17037178205?pr=1724

**Review by Kobzol — COMMENTED** [review]
Thanks for the changes. Some problems that cause CI to fail still have to be resolved, as per the other comments.

Also, as per https://github.com/rust-lang/rustc-perf/pull/1724#discussion_r1332925816, we shouldn't include `Clippy` in `Profile::all()` for now. Let's just remove it from there, that should fix collection on the server. Later we might add it back once we're sure that clippy benchmarks can be run on the perf server.

File: collector/src/toolchain.rs (hunk 10 above)
Comment (blyxyas) [reply collector/src/toolchain.rs]: Done :cat:!

File: collector/src/toolchain.rs (hunk 9 above)
Comment (blyxyas) [reply collector/src/toolchain.rs]: Probably done, I'll check CI to see if the fix I did is correct :cat:

File: collector/src/toolchain.rs [hunk 11]
```diff
@@ -123,7 +123,10 @@ impl SysrootDownload {
         let components = ToolchainComponents::from_binaries_and_libdir(
             sysroot_bin("rustc")?,
             Some(sysroot_bin("rustdoc")?),
-            Some(sysroot_bin("clippy")?),
+            match sysroot_bin("cargo-clippy") {
```
Comment (Kobzol) [inline collector/src/toolchain.rs:126]: There is a helper `or` method for this (https://doc.rust-lang.org/std/result/enum.Result.html#method.ok) in the stdlib, btw :) But you can keep it as it is here.

**Review by Kobzol — APPROVED** [review]
Great, thanks for the changes! :)

File: collector/src/toolchain.rs (hunk 11 above)
Comment (blyxyas) [reply collector/src/toolchain.rs:126]: Do I... *not know `std`* :fearful:. Haha I'll change it right now.
